Mathlib/Combinatorics/Quiver/ConnectedComponent.lean

```diff
@@ -50,18 +57,149 @@ instance [Inhabited V] : Inhabited (WeaklyConnectedComponent V) :=
+   fun _ => ⟨⟨Path.nil⟩, ⟨Path.nil⟩⟩,
+   fun ⟨hab, hba⟩ => ⟨hba, hab⟩,
+   fun ⟨hab, hba⟩ ⟨hbc, hcb⟩ =>
+     ⟨⟨hab.some.comp hbc.some⟩, ⟨hcb.some.comp hba.some⟩⟩⟩
+
+/-- The type of strongly connected components (bidirectional reachability classes). -/
+def StronglyConnectedComponent (V : Type*) [Quiver V] : Type _ :=
+  Quotient (stronglyConnectedSetoid V)
+
+namespace StronglyConnectedComponent
+
+variable {V}
+
+/-- The canonical map from a vertex to its strongly connected component. -/
+protected def mk : V → StronglyConnectedComponent V :=
+  @Quotient.mk' _ (stronglyConnectedSetoid V)
+
+instance : Coe V (StronglyConnectedComponent V) :=
+  ⟨StronglyConnectedComponent.mk⟩
+
+instance [Inhabited V] : Inhabited (StronglyConnectedComponent V) :=
+  ⟨(default : V)⟩
+
+protected lemma eq (a b : V) :
+  (a : StronglyConnectedComponent V) = b
+    ↔ (Nonempty (Path a b) ∧ Nonempty (Path b a)) := Quotient.eq''
+
+@[simp] lemma mk_eq_mk {a b : V} :
+    (StronglyConnectedComponent.mk a : StronglyConnectedComponent V) =
+    StronglyConnectedComponent.mk b ↔
+    (Nonempty (Path a b) ∧ Nonempty (Path b a)) :=
+  StronglyConnectedComponent.eq a b
+
+lemma IsSStronglyConnected.pos_cycle
+    {V} [Quiver V] (h : IsSStronglyConnected V) (v : V) :
+    ∃ p : Path v v, 0 < p.length := h v v
+
+end StronglyConnectedComponent
+
+variable {V}
+
+lemma stronglyConnectedComponent_eq_of_path {a b : V}
+    (hab : Nonempty (Path a b)) (hba : Nonempty (Path b a)) :
+    (a : StronglyConnectedComponent V) = b :=
+  (StronglyConnectedComponent.eq (a := a) (b := b)).2 ⟨hab, hba⟩
+
+lemma exists_path_of_stronglyConnectedComponent_eq {a b : V}
+    (h : (a : StronglyConnectedComponent V) = b) :
+    (Nonempty (Path a b)) ∧ (Nonempty (Path b a)) :=
+  (StronglyConnectedComponent.eq (a := a) (b := b)).1 h
+
+lemma stronglyConnectedComponent_singleton_iff {V} [Quiver V] (v : V) :
+    (∀ w : V, (w : StronglyConnectedComponent V) = v → w = v) ↔
+    (∀ w : V, w ≠ v → ¬(Nonempty (Path v w) ∧ Nonempty (Path w v))) := by
+  constructor
+  · intro h_singleton w hw_ne h_bidir
+    obtain ⟨hab, hba⟩ := h_bidir
+    have h_same_scc : (w : StronglyConnectedComponent V) = v :=
+      stronglyConnectedComponent_eq_of_path (a := w) (b := v) hba hab
+    obtain rfl := h_singleton w h_same_scc
+    exact hw_ne rfl
+  · intro h_no_bidir w h_same_scc
+    by_contra hw_ne
+    obtain ⟨hab, hba⟩ :=
+      exists_path_of_stronglyConnectedComponent_eq (a := w) (b := v) h_same_scc
+    exact (h_no_bidir w hw_ne) ⟨hba, hab⟩
+
+lemma symmetrify_isStronglyConnected_of_isStronglyConnected {V} [Quiver V]
+    (h : IsStronglyConnected V) : IsStronglyConnected (Symmetrify V) := by
+  intro a b
+  obtain ⟨p⟩ := h a b
+  induction p with
+  | nil => exact ⟨Path.nil⟩
+  | cons q e ih => exact ⟨ih.some.cons (Sum.inl e)⟩
+
+lemma isSStronglyConnected_of_hasEdge {V} [Quiver V]
+    (h_sc : IsStronglyConnected V)
+    (h_edge : ∃ (i j : V), Nonempty (i ⟶ j)) :
```

Comment:
It's almost always better to destruct an existential in an argument to save the caller and proof bundling / unbundling:
```suggestion
    {i₀ j₀ : V} (e₀ : i₀ ⟶ j₀) :
```